File: Java/src/Convert Sorted List to Binary Search Tree.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; next = null; }
 * }
 */
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    private ListNode node;
    public TreeNode sortedListToBST(ListNode head) {
        int length = 0;
        if (head == null) return null;
        ListNode cur = head;
        node = head;
        while (cur != null){
            length++ ;
            cur = cur.next;
        }
        return helper(0, length-1);
    }
    private TreeNode helper(int start, int end){
        if (start > end) return null;
        int mid = (start+end)/2;
        TreeNode left = helper(start, mid-1);
        TreeNode root = new TreeNode(node.val);
        root.left = left;
        node = node.next;
        TreeNode right = helper(mid+1, end);
        root.right = right;
        return root;
    }
}

File: Python/Convert Sorted List to Binary Search Tree .py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
#
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a list node
    # @return a tree node
    def sortedListToBST(self, head):
        if not head:
            return None
    
        return self.generate(head, None)
            
    def generate(self, head, tail):
        if not head:
            return None
        
        first = head
        second = head
        
        while second != tail and second.next != tail:
            first = first.next
            second = second.next.next
        
        node = TreeNode(first.val)
        node.left = self.generate( None if head == first else head, first)
        node.right = self.generate( None if first.next == tail else first.next, tail)
        return node
        
        
        
        